### PR TITLE
MM-19096 – simplify custom login process

### PR DESCRIPTION
**Pull request by deanwhillier**
**Summary**
The new functionality added to support 3rd party custom login mechanisms (SAML, OAUTH) following a security enhancement attempted to determine the 3rd party hostname and temporarily trust it for the duration of the login request. This temporary trust created some challenges during testing and it was subsequently decided that this aspect of the feature was too brittle given we ultimately can't control what an IDP does on their end to validate an authentication request, nor can we control the kinds of linkable resources an end user might be presented with during the login process.

This PR removes the temporary trust aspect of the feature and solely relies on appropriately flagging a custom login is in progress, during which any navigation that needs to occur until the user is returned to MM will be allowed to occur. Once the login process completes, the flag is reset and normal navigation handling resumes.

As a by-product of the change, this PR now allows links on IDP pages to function as intended, fixing the issue [reported here](https://mattermost.atlassian.net/browse/MM-18936).

**Issue links**
Primary: https://mattermost.atlassian.net/browse/MM-19096
Secondary: https://mattermost.atlassian.net/browse/MM-18936
Original feature: https://mattermost.atlassian.net/browse/MM-17255

### CODE REVIEW

File: src/main.js [hunk 1]
```diff
@@ -763,40 +732,44 @@ function handleMainWindowWebContentsCrashed() {
...
-function isTrustedURL(url) {
+function parseURL(url) {
   if (!url) {
-    return false;
+    return null;
   }
-
-  let parsedUrl = url;
+  let parsedURL = url;
   if (typeof url === 'string') {
```
Comment (lieut-data) [inline src/main.js]: nit: if `url` isn't a string, we end up returning something that might not be an `URL`. Should we test for an `URL` and return that explicitly, but otherwise return `null`?

File: src/main.js [hunk 2]
```diff
@@ -763,40 +732,44 @@ function handleMainWindowWebContentsCrashed() {
...
-
+function isTrustedURL(url) {
+  const parsedURL = parseURL(url);
+  if (!parsedURL) {
+    return false;
+  }
+  const trustedURLs = config.teams.map((team) => parseURL(team.url));
   for (const trustedURL of trustedURLs) {
-    if (parsedUrl.origin === trustedURL.origin) {
+    if (parsedURL.origin === trustedURL.origin) {
```
Comment (lieut-data) [inline src/main.js]: Might `trustedURL` be `null` here based on the `parseURL` semantics above?

File: src/main.js [hunk 3]
```diff
@@ -383,60 +381,31 @@ function handleAppWebContentsCreated(dc, contents) {
...
+  //    - indicate custom login is in progress
+  // 2. are we finished with the custom login process?
+  //    - indicate custom login is NOT in progress
   contents.on('did-start-navigation', (event, url) => {
     const contentID = event.sender.id;
-    const parsedUrl = new URL(url);
-    const urlIsTrusted = isTrustedURL(parsedUrl);
-    const urlIsCustomLoginPath = isCustomLoginURL(parsedUrl);
+    const parsedURL = parseURL(url);
+    const urlIsTrusted = isTrustedURL(parsedURL);
```
Comment (lieut-data) [inline src/main.js]: nit: maybe early exit if `!urlIsTrusted`?

File: src/main.js [hunk 4]
```diff
@@ -383,60 +381,31 @@ function handleAppWebContentsCreated(dc, contents) {
...
+    const urlIsCustomLoginPath = isCustomLoginURL(parsedURL);
     const previousPage = event.sender.history[event.sender.history.length - 1];
 
     if (urlIsTrusted && urlIsCustomLoginPath && !customLogins[contentID].inProgress && previousPage.endsWith('/login')) {
```
Comment (lieut-data) [inline src/main.js]: Will this `endsWith` check fail if there are query string parameters? Is it strictly necessary to check `previousPage` at all?

File: src/main.js [hunk 5]
```diff
@@ -383,60 +381,31 @@ function handleAppWebContentsCreated(dc, contents) {
 
   contents.on('will-navigate', (event, url) => {
     const contentID = event.sender.id;
-    const parsedUrl = new URL(url);
-    const urlIsTrusted = isTrustedURL(parsedUrl);
-    const urlIsTrustedExternalLoginPath = parsedUrl.hostname === customLogins[contentID].externalHostname;
+    const parsedURL = parseURL(url);
+    const urlIsTrusted = isTrustedURL(parsedURL);
 
     // don't prevent custom login attempts (oath, saml)
```
Comment (lieut-data) [inline src/main.js]: nit: the double negatives here are hard to parse. Perhaps something like:
```js
    if (urlIsTrusted) {
        return;
    }

    // allow arbitrary navigation while in a trusted login flow
    if (customLogins[contentID].inProgress) {
       return;
    }

    // Otherwise, prevent the webview from navigating outside the trusted domain. This
    // prevents security issues such as...
    event.preventDefault();
```

File: src/main.js (hunk 4 above)
Comment (deanwhillier) [reply src/main.js]: Good question. @crspeller, do you think it's necessary to ensure a SAML/OAUTH endpoint is accessed directly from the /login page in order to trigger the updated custom login process?

File: src/main.js (hunk 5 above)
Comment (deanwhillier) [reply src/main.js]: I like this approach, just have to keep reminding myself to break my old habits.

File: src/main.js (hunk 1 above)
Comment (deanwhillier) [reply src/main.js]: Reworking this a bit ...

**Comment by deanwhillier** [comment]
Pushed some updates, left the logic to check for `/login` for now, pending @crspeller's thoughts.

File: src/main.js (hunk 1 above)
Comment (deanwhillier) [reply src/main.js]: So the update to this one returns the url if it is an instance of URL and exits. If not, it tries to convert it to a URL and returns that if successful, otherwise returns null.

File: src/main.js [hunk 6]
```diff
@@ -383,60 +381,38 @@ function handleAppWebContentsCreated(dc, contents) {
...
-        // redirect to starting url if set
-        if (customLogins[contentID].startingURL) {
-          event.sender.loadURL(customLogins[contentID].startingURL);
-          customLogins[contentID].startingURL = null;
-        }
-      }
+    const parsedURL = parseURL(url);
+    const urlIsTrusted = isTrustedURL(parsedURL);
+
+    if (urlIsTrusted) {
```
Comment (crspeller) [inline src/main.js]:
```suggestion
    if (isTrustedURL(parsedURL)) {
```

Comment (crspeller) [reply src/main.js]: This suggestion can be applied in a bunch of places here. I think it's leftovers from when this was more complicated.

File: src/main.js (hunk 4 above)
Comment (crspeller) [reply src/main.js]: Probably not necessary. This can be removed. Worried about some kind of deep login link failing because of this.

Comment (deanwhillier) [reply src/main.js]: K, sounds good, thanks!
